### ISSUE TITLE
Implement em::findOneOrFail

### ISSUE THREAD

**Issue by kgoncharov** [mikro-orm/mikro-orm#133]
**Is your feature request related to a problem? Please describe.**
Sometimes it's very convenient to have a short method that will fail-fast if nothing was found.

**Describe the solution you'd like**
`em::findOrFail` that return `Promise<T>` or throw exception if entity wasn't found.

**Comment by B4nan** [comment]
Agreed, I am usually defining this method in custom base repository, but sure we can have it integrated as well. Personally I am using custom `NotFoundException` that is later on catched in main request handler and handled as 404 response. I guess we could allow configuring a callback where users could provide their own exception and possibly error message based on what was not found.

**Comment by kgoncharov** [comment]
I supposed that this method would be used only for situations, where the entity should be in database, and if there is no entity server should return `InternalServerException` at the end, to show that there's some inconsistency in database.
For this purposes `findOneOrFail` name would be the best.
On the other hand there could be two methods:
`findOneOrThrow` - for custom handling on what exception we're going to throw (`NotFoundException`, etc)
`findOneOrFail` - for fatal and critical errors that will return some critical custom exception that could be propagated to `InternalServerException`
@B4nan wdyt?

**Comment by B4nan** [comment]
I think that could cause a lot of confusion, I would like to have only one method for this. 

Btw for your `InternalServerException` use case, how are you handling 404s? I guess manually, right? Then with my proposal you would just set up the callback to produce `InternalServerException` and keep handling the 404 manually (either via try/catch or via truthy checks). Or as I already said, you can always define your own methods in base repository (you can set your own base repository for all entities in ORM's configuration).

**Comment by kgoncharov** [comment]
Yep, manually.
Anyway, just `findOneOrFail` with capability to set custom exception to throw would be great.

**Comment by vimmerru** [comment]
I would like 

```
cons used = findOneOrFail(id, () => InternalServerError(`No expected user in db: ${id}`));
```

**Comment by B4nan** [comment]
findOne already has more complex interface:

    async findOne<T extends IEntityType<T>>(entityName: EntityName<T>, where: FilterQuery<T> | IPrimaryKey, populate?: string[] | FindOneOptions, orderBy?: QueryOrderMap): Promise<T | null> {

But yes, I like the idea to allow overriding the exception handler via it's parameters, probably via `FindOneOptions` (probably `FindOneOrFailOptions` that will extend that).

**Comment by B4nan** [comment]
Available for testing in `3.0.0-alpha.9`. Closing as resolved. 

Docs: https://github.com/mikro-orm/mikro-orm/commit/0d57b7b3320e3292247f6919ac1a336dc1981afb#diff-474106f26d234ca3b44599c73da3d296R112
